core/arch/arm/sm/suspend.S

```diff
@@ -0,0 +1,244 @@
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <kernel/unwind.h>
+#include <asm.S>
+#include <asm-defines.h>
+#include <arm.h>
+#include <arm32_macros.S>
+#include <arm32_macros_cortex_a9.S>
+#include <kernel/asan.h>
+#include <platform_config.h>
+#include <keep.h>
+#include <kernel/unwind.h>
+
+#define PRRR			0x034
+#define NMRR			0x038
+
+.section .text
+FUNC cpu_suspend, :
+UNWIND( .fnstart)
+	push	{r4 - r12, lr}
+UNWIND(.save	{r4 - r12, lr})
+	mov	r5, sp
+	sub	sp, sp, #PM_CTX_SIZE
+	ldr	r3, =suspend_sp
+	read_mpidr r6
+	/*
+	 * Note:
+	 * We assume max 4 cpus and only 1 cluster.
+	 * each cpu consumes 2 words in suspend_sp space
+	 * the suspend_sp->virt = ((u32)suspend_sp + 4 * cpu_id)
+	 */
+	and	r6, r6, #0x3
+	lsl	r6, r6, #0x3
+	push	{r0, r1}
+	add	r3, r3, r6
+
+	mov	r2, r5
+	mov	r1, #PM_CTX_SIZE
+	add	r0, sp, #8
+	blx	cpu_suspend_save
+	adr	lr, cpu_suspend_panic
+	/* Jump to arch specific suspend*/
+	pop	{r0, pc}
+	/* cpu not suspended */
+	add	sp, sp, #PM_CTX_SIZE
+	pop	{r4 - r12, lr}
+	/* Return -1 to the caller */
+	mov	r0, #(-1)
+	bx	lr
+UNWIND( .fnend)
+END_FUNC cpu_suspend
+
+FUNC cpu_suspend_panic, :
+UNWIND( .fnstart)
+	b	.
+UNWIND( .fnend)
+END_FUNC cpu_suspend_panic
+
+FUNC cpu_v7_suspend, :
+UNWIND(.fnstart)
+	push	{r4 - r12, lr}
+	read_fcseidr r4
+	read_tpidruro r5
+	stmia	r0!, {r4 - r5}
+	read_dacr r4
+#ifdef CFG_WITH_LPAE
+#error "Not supported"
+#else
+	read_ttbr0 r5
+	read_ttbr1 r6
+	read_ttbcr r7
+#endif
+	read_sctlr r8
+	read_actlr r9
+	read_cpacr r10
+	read_mvbar r11
+	stmia	r0!, {r4 - r11}
+	read_prrr r4
+	read_nmrr r5
+	read_vbar r6
+	read_nsacr r7
+	stmia	r0, {r4 - r7}
+	pop	{r4 - r12, pc}
+UNWIND( .fnend)
+END_FUNC cpu_v7_suspend
+
+FUNC cpu_do_suspend, :
+UNWIND(.fnstart)
+	push	{r4, r5}
+	read_midr r4
+	ubfx	r5, r4, #4, #12
+	ldr	r4, =0xc07
```

Comment:
This magic number could be a defined in `<arm32.h>` instead.
The assignment could also use `mov_imm` instead.